pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -51,9 +52,19 @@ func NewController() *Controller {
 
 // AddRegistry adds registries into the aggregated controller
 func (c *Controller) AddRegistry(registry Registry) {
+	c.storeLock.Lock()
+	defer c.storeLock.Unlock()
 	c.registries = append(c.registries, registry)
 }
 
+// GetRegistries safely gets a slice of registries
+func (c *Controller) GetRegistries() []Registry {
+	c.storeLock.RLock()
+	defer c.storeLock.RUnlock()
+	registries := c.registries
+	return registries
+}
```

Comment:
I don't think that this will work. When the delete is implemented, it'd run into a situation where a query is made to a registry that's just deleted. 

The purpose of Read write lock is that it'll be fine for 99 percent of the situation where everyone is reading so that they can be done concurrently.

Reply:
As per Slack chat, I have filed two issues to followup.
https://github.com/istio/istio/issues/5605
https://github.com/istio/istio/issues/5606